test/e2e/cluster_upgrade_runtimesdk.go

```diff
@@ -281,6 +286,71 @@ func clusterUpgradeWithRuntimeSDKSpec(ctx context.Context, inputGetter func() cl
 	})
 }
 
+// machineSetPreflightChecksTestHandler verifies the MachineSet preflight checks.
+// At this point in the test the ControlPlane is upgraded to the new version and the upgrade to the MachineDeployments
+// should be blocked by the AfterControlPlaneUpgrade hook.
+// Test the MachineSet preflight checks by scaling up the MachineDeployment. The creation on the new Machine
+// should be blocked because the preflight checks should not pass (kubeadm version skew preflight check should fail).
+func machineSetPreflightChecksTestHandler(ctx context.Context, c client.Client, clusterRef types.NamespacedName) {
+	// Verify that the hook is called and the topology reconciliation is blocked.
+	hookName := "AfterControlPlaneUpgrade"
+	Eventually(func() error {
+		if err := checkLifecycleHooksCalledAtLeastOnce(ctx, c, clusterRef, []string{hookName}); err != nil {
+			return err
+		}
+
+		cluster := framework.GetClusterByName(ctx, framework.GetClusterByNameInput{
+			Name: clusterRef.Name, Namespace: clusterRef.Namespace, Getter: c})
+
+		if !clusterConditionShowsHookBlocking(cluster, hookName) {
+			return errors.Errorf("Blocking condition for %s not found on Cluster object", hookName)
+		}
+
+		return nil
+	}, 30*time.Second).Should(Succeed(), "%s has not been called", hookName)
+
+	// Scale up the MachineDeployment
+	machineDeployments := framework.GetMachineDeploymentsByCluster(ctx, framework.GetMachineDeploymentsByClusterInput{
+		Lister:      c,
+		ClusterName: clusterRef.Name,
+		Namespace:   clusterRef.Namespace,
+	})
+	md := machineDeployments[0]
+
+	patchHelper, err := patch.NewHelper(md, c)
+	Expect(err).To(BeNil())
+
+	// Scale up the MachineDeployment.
+	// Since the MachineDeployment is pending an upgrade at this point the topology controller will not push any changes
+	// to the MachineDeployment. Therefore, the changes made to the MachineDeployment here will not be replaced
+	// immediately.
+	*md.Spec.Replicas += 1
+	Eventually(func() error {
+		return patchHelper.Patch(ctx, md)
+	}).Should(Succeed(), "Failed to scale up the MachineDeployment %s", klog.KObj(md))
+
+	// Verify that the MachineSet is not creating the new Machine.
+	Eventually(func(g Gomega) {
+		machineSets := framework.GetMachineSetsByCluster(ctx, framework.GetMachineSetsByClusterInput{
```

Comment:
Wondering if we should get the 1 MS of the MD we scaled up? Currently this all relies on that we only have one MD and one MS in the cluster, which might be a bit brittle